File: zy70055/src/services/exceptionService.ts
import { Prisma } from '@prisma/client';
import { prisma } from '../config';
import { ExceptionType } from '../constants';
import { ResourceNotFoundError } from '../utils/error';
import { generateExceptionNo, generateOperationNo } from '../utils/generator';
import { logger } from '../utils/logger';

interface CreateExceptionParams {
  batchId: string;
  exceptionType: string;
  severity: 'HIGH' | 'MEDIUM' | 'LOW';
  message: string;
  detail?: string;
  operator?: string;
  ipAddress?: string;
}

class ExceptionService {
  async createException(params: CreateExceptionParams) {
    const exceptionNo = generateExceptionNo();

    const exception = await prisma.exceptionRecord.create({
      data: {
        exceptionNo,
        batchId: params.batchId,
        exceptionType: params.exceptionType,
        severity: params.severity,
        message: params.message,
        detail: params.detail,
      },
    });

    await prisma.operationLog.create({
      data: {
        operationNo: generateOperationNo(),
        batchId: params.batchId,
        operator: params.operator || 'system',
        operationType: 'CREATE_EXCEPTION',
        detail: `创建异常记录: ${exceptionNo} - ${params.message}`,
        ipAddress: params.ipAddress,
      },
    });

    logger.warn(`异常记录创建: ${exceptionNo}`, { batchId: params.batchId, type: params.exceptionType });
    return exception;
  }

  async getExceptionsByBatch(batchId: string, includeResolved: boolean = false) {
    const where: Prisma.ExceptionRecordWhereInput = { batchId };
    if (!includeResolved) {
      where.isResolved = false;
    }

    return prisma.exceptionRecord.findMany({
      where,
      orderBy: { createdAt: 'desc' },
    });
  }

  async getPendingExceptions() {
    return prisma.exceptionRecord.findMany({
      where: { isResolved: false },
      include: {
        batch: {
          include: { merchant: true },
        },
      },
      orderBy: { createdAt: 'desc' },
    });
  }

  async resolveException(
    exceptionNo: string,
    resolvedBy: string,
    resolutionNote: string
  ) {
    const exception = await prisma.exceptionRecord.findUnique({
      where: { exceptionNo },
    });

    if (!exception) {
      throw new ResourceNotFoundError(`异常记录不存在: ${exceptionNo}`);
    }

    if (exception.isResolved) {
      return { message: '该异常已处于已解决状态', exception };
    }

    const resolved = await prisma.exceptionRecord.update({
      where: { exceptionNo },
      data: {
        isResolved: true,
        resolvedAt: new Date(),
        resolvedBy,
        resolutionNote,
      },
    });

    await prisma.operationLog.create({
      data: {
        operationNo: generateOperationNo(),
        batchId: exception.batchId,
        operator: resolvedBy,
        operationType: 'RESOLVE_EXCEPTION',
        detail: `解决异常: ${exceptionNo} - ${resolutionNote}`,
      },
    });

    logger.info(`异常已解决: ${exceptionNo}`, { resolvedBy });
    return resolved;
  }

  async getBatchExceptionSummary(batchId: string) {
    const exceptions = await prisma.exceptionRecord.findMany({
      where: { batchId },
    });

    const pending = exceptions.filter(e => !e.isResolved);
    const resolved = exceptions.filter(e => e.isResolved);

    const byType: Record<string, number> = {};
    for (const ex of exceptions) {
      byType[ex.exceptionType] = (byType[ex.exceptionType] || 0) + 1;
    }

    const bySeverity: Record<string, number> = {};
    for (const ex of exceptions) {
      bySeverity[ex.severity] = (bySeverity[ex.severity] || 0) + 1;
    }

    return {
      total: exceptions.length,
      pending: pending.length,
      resolved: resolved.length,
      byType,
      bySeverity,
      pendingTypes: [...new Set(pending.map(e => e.exceptionType))],
    };
  }

  async batchResolveExceptions(
    exceptionNos: string[],
    resolvedBy: string,
    resolutionNote: string
  ) {
    const results: { exceptionNo: string; success: boolean; message?: string }[] = [];

    for (const no of exceptionNos) {
      try {
        await this.resolveException(no, resolvedBy, resolutionNote);
        results.push({ exceptionNo: no, success: true });
      } catch (e: any) {
        results.push({ exceptionNo: no, success: false, message: e.message });
      }
    }

    return results;
  }
}

export const exceptionService = new ExceptionService();

File: zy70191/app/services/exception_service.py
from typing import List, Optional
from datetime import datetime
from sqlalchemy.orm import Session

from app.models import ExceptionRecord, ExceptionType
from app.schemas import ExceptionRecordCreate


class ExceptionService:
    def __init__(self, db: Session):
        self.db = db

    def record_exception(
        self,
        exception_type: ExceptionType,
        description: str,
        detail: Optional[str] = None,
        switch_request_id: Optional[int] = None
    ) -> ExceptionRecord:
        exception = ExceptionRecord(
            switch_request_id=switch_request_id,
            exception_type=exception_type,
            description=description,
            detail=detail
        )
        self.db.add(exception)
        self.db.commit()
        self.db.refresh(exception)
        return exception

    def get_exception(self, exception_id: int) -> Optional[ExceptionRecord]:
        return self.db.query(ExceptionRecord).filter(ExceptionRecord.id == exception_id).first()

    def get_all_exceptions(
        self,
        is_resolved: Optional[bool] = None,
        exception_type: Optional[ExceptionType] = None
    ) -> List[ExceptionRecord]:
        query = self.db.query(ExceptionRecord)
        
        if is_resolved is not None:
            query = query.filter(ExceptionRecord.is_resolved == is_resolved)
        if exception_type:
            query = query.filter(ExceptionRecord.exception_type == exception_type)
        
        return query.order_by(ExceptionRecord.created_at.desc()).all()

    def get_pending_exceptions(self) -> List[ExceptionRecord]:
        return self.get_all_exceptions(is_resolved=False)

    def resolve_exception(
        self,
        exception_id: int,
        resolved_by: str,
        resolution_detail: Optional[str] = None
    ) -> Optional[ExceptionRecord]:
        exception = self.get_exception(exception_id)
        if exception:
            exception.is_resolved = True
            exception.resolved_by = resolved_by
            exception.resolved_at = datetime.utcnow()
            if resolution_detail:
                if exception.detail:
                    exception.detail = f"{exception.detail}\n\n处理说明: {resolution_detail}"
                else:
                    exception.detail = f"处理说明: {resolution_detail}"
            self.db.commit()
            self.db.refresh(exception)
        return exception

    def get_exceptions_by_request(self, switch_request_id: int) -> List[ExceptionRecord]:
        return self.db.query(ExceptionRecord).filter(
            ExceptionRecord.switch_request_id == switch_request_id
        ).order_by(ExceptionRecord.created_at.desc()).all()

    def get_exception_statistics(self) -> dict:
        total = self.db.query(ExceptionRecord).count()
        pending = self.db.query(ExceptionRecord).filter(ExceptionRecord.is_resolved == False).count()
        resolved = total - pending
        
        by_type = {}
        for et in ExceptionType:
            count = self.db.query(ExceptionRecord).filter(
                ExceptionRecord.exception_type == et
            ).count()
            by_type[et.value] = count

        return {
            "total": total,
            "pending": pending,
            "resolved": resolved,
            "by_type": by_type
        }
